clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2576,6 +2678,17 @@ bool Util::isSyclStreamType(const QualType &Ty) {
   return isSyclType(Ty, "stream");
 }
 
+bool Util::isSyclHalfType(const QualType &Ty) {
+  const StringRef &Name = "half";
+  std::array<DeclContextDesc, 5> Scopes = {
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "detail"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "half_impl"},
+      Util::DeclContextDesc{Decl::Kind::CXXRecord, Name}};
```

Comment:
Please, add an attribute to annotate SYCL types like accessor, sampler, image. It's much more reliable approach than declaration name, which introduces hard-coded dependency on exact symbol name in the runtime library.
I suppose we need only one attribute to detect special types which must be handled as separate kernel parameters. If not, we can parameterize the attribute.

This refactoring can be done in a separate PR.
Tagging @Naghasan, who was going to contribute similar attribute from ComputeCPP implementation.

Reply:
> Please, add an attribute to annotate SYCL types like accessor, sampler, image. It's much more reliable approach than declaration name, which introduces hard-coded dependency on exact symbol name in the runtime library.
> I suppose we need only one attribute to detect special types which must be handled as separate kernel parameters. If not, we can parameterize the attribute.
> 
> This refactoring can be done in a separate PR.
> Tagging @Naghasan, who was going to contribute similar attribute from ComputeCPP implementation.

Right now just one attribute is not enough since we have differences between handling of different classes, especially in integration header support. It seems runtime library needs to handle each type in a different way, I have no idea how to redo it so it will be unified across different SYCL types.
Right now we can start with parameterized attribute, i.e. for example:
```
class accessor {
...
__attribute__((sycl_special_type(accessor))) DataT *Data;
} 

...
class sampler {
...
__attribute__((sycl_special_type(sampler))) ocl_sampler_t Sampler;
} 
```
Also, we can discuss it in upcoming upstreaming meeting.
I've created an issue to keep track on that https://github.com/intel/llvm/issues/2041.

Reply:
I'd like to point out that clang is able to differentiate sampler from accessor w/o any attributes - they are different **types** (pointer type vs __ocl_sampler_t).

Reply:
Okay, makes sense.